pkg/ruler/ruler_test.go

```diff
@@ -216,3 +227,324 @@ func compareRuleGroupDescToStateDesc(t *testing.T, expected *rules.RuleGroupDesc
 		require.Equal(t, expected.Rules[i].Alert, got.ActiveRules[i].Rule.Alert)
 	}
 }
+
+func TestSharding(t *testing.T) {
+	const (
+		user1 = "user1"
+		user2 = "user2"
+		user3 = "user3"
+	)
+
+	user1Group1 := &rules.RuleGroupDesc{User: user1, Namespace: "namespace", Name: "first"}
+	user1Group2 := &rules.RuleGroupDesc{User: user1, Namespace: "namespace", Name: "second"}
+	user2Group1 := &rules.RuleGroupDesc{User: user2, Namespace: "namespace", Name: "first"}
+	user3Group1 := &rules.RuleGroupDesc{User: user3, Namespace: "namespace", Name: "first"}
+
+	// Must be distinct for test to work.
+	user1Group1Token := tokenForGroup(user1Group1)
+	user1Group2Token := tokenForGroup(user1Group2)
+	user2Group1Token := tokenForGroup(user2Group1)
+	user3Group1Token := tokenForGroup(user3Group1)
+
+	allRules := map[string]rules.RuleGroupList{
+		user1: {user1Group1, user1Group2},
+		user2: {user2Group1},
+		user3: {user3Group1},
+	}
+
+	type testCase struct {
+		sharding         bool
+		shardingStrategy string
+		shuffleShardSize int
+		setupRing        func(*ring.Desc)
+
+		expectedRulesForFuler1 map[string]rules.RuleGroupList
```

Comment:
`Fuler` made me smile :D Thanks!

```suggestion
		expectedRulesForRuler1 map[string]rules.RuleGroupList
```

Reply:
> `Fuler` made me smile :D Thanks!

I hope it doesn't mean anything bad :)